### PR TITLE
Convert comp/core/config to module

### PR DESCRIPTION
**Pull request by dineshg13**
### What does this PR do?

This PR creates a `comp/core/config` module. In the process, we do the following:

- Convert `cmd/agent/common/path` into a module 
- convert `pkg/util/hostname/validate` into module
- Move code from `pkg/config` to new package/module `pkg/config/setup.` pkg/config/setup` will hold all the code in initializing the env vars for DD products and loading config from files. 

### Motivation
[RFC](https://docs.google.com/document/d/1uKWa9rMNm0U-euEzEJ69KW7Id5u1E4GfYHyzC1SzYoI/edit#heading=h.wnkz5u7vx1fb)

### Describe how to test/QA your changes

Agent should be able to read config both via Envars and config file in all environments. 
-  Install agent via helm to verify in containerized environment. 
-  Run agent in bare metal and verify we are able to receiver metrics/logs and traces.  
-  Run agent in windows and verify the config location and we are able to receive metrics/traces/logs. 
-  For process-agent, ensure that process agent status output still works correctly. The command to check the process agent status is: `/opt/datadog-agent/embedded/bin/process-agent status`

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by brycekahle** [comment]
Can you update CODEOWNERS for all the files that were moved?

File: pkg/config/aliases_darwin.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
```
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/config/aliases_darwin.go:5]: I don't see go build in `https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_darwin.go`

File: pkg/config/aliases_windows.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
```
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/config/aliases_windows.go:5]: I don't see go build in `https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_windows.go`

[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/config/aliases_windows.go:5]: It doesn't need one, the `_windows` suffix is an implied build tag. From https://pkg.go.dev/cmd/go#hdr-Build_constraints

> If a file's name, after stripping the extension and a possible _test suffix, matches any of the following patterns:
> *_GOOS
> *_GOARCH
> *_GOOS_GOARCH
> (example: source_windows_amd64.go) where GOOS and GOARCH represent any known operating system and architecture values respectively, then the file is considered to have an implicit build constraint requiring those terms (in addition to any explicit constraints in the file).

**Review by chouetz — APPROVED** [review]
Ok for agent-platform files

File: pkg/config/setup_test.go
Comment (pgimalac) [inline pkg/config/setup_test.go:1]: Where does this come from ?
I think moving it to the component directory would avoid duplicating the `testdata/full.yaml` file ?

File: pkg/config/setup/test_helpers.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+// confFromYAML generates a configuration from the given yaml config
+func confFromYAML(yamlConfig string) pkgconfigmodel.Config {
+	conf := conf()
+	conf.SetConfigType("yaml")
+	e := conf.ReadConfig(bytes.NewBuffer([]byte(yamlConfig)))
+	if e != nil {
+		log.Println(e)
+	}
+	return conf
+}
```
Comment (pgimalac) [inline pkg/config/setup/test_helpers.go:32]: Is this different from the functions in `pkg/config/test_helpers.go` ?
If you need them from `setup`, can you actually move them and just change the import path in `pkg/config` ? (or at worst add an alias, but avoiding duplication)

**Review by pgimalac — CHANGES_REQUESTED** [review]
Most comments are nitpicks, a few are actual blockers as I believe they might introduce some bugs.
Reviewing this huge PR which moves things around is quite hard honestly, it would be way simpler/faster if you were able to split it into several smaller PRs.

File: pkg/config/setup/process_test.go
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/config/setup/process_test.go:1]: Yes. It is becoz of lint issue.

File: pkg/config/aliases.go [hunk 4]
```diff
@@ -113,3 +118,138 @@ func GetSyslogURI() string {
...
+
+// Datatype Aliases
+const (
+	Metrics = pkgconfigsetup.Metrics
+	Traces  = pkgconfigsetup.Traces
+	Logs    = pkgconfigsetup.Logs
+)
+
+// Aliases for config defaults
+const (
```
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/config/aliases.go:140]: yes. Not a fan as well. But it is a patch that would get us across the line. Cleaning up is an easy PR.

File: pkg/config/setup/config.go [hunk 5]
```diff
@@ -198,16 +193,10 @@ const (
 
 func init() {
 	osinit()
-	// Configure Datadog global configuration
-	Datadog = NewConfig("datadog", "DD", strings.NewReplacer(".", "_"))
-	SystemProbe = NewConfig("system-probe", "DD", strings.NewReplacer(".", "_"))
-	// Configuration defaults
-	InitConfig(Datadog)
-	InitSystemProbeConfig(SystemProbe)
```
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/config/setup/config.go:206]: Good catch. Am reverting them

File: pkg/config/setup_test.go
Comment (dineshg13) [reply pkg/config/setup_test.go:1]: Reverted

File: pkg/config/aliases_darwin.go
[the remark this answers is not in the document]
Comment (dineshg13) [reply pkg/config/aliases_darwin.go:1]: They alias different variables. For example, windows one defines them as var while others define them as const.

File: pkg/config/setup/config.go [hunk 6]
```diff
@@ -20,18 +20,21 @@ import (
...
 	pkgconfigmodel "github.com/DataDog/datadog-agent/pkg/config/model"
-	"github.com/DataDog/datadog-agent/pkg/util/optional"
-
 	"github.com/DataDog/datadog-agent/pkg/util/hostname/validate"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/optional"
+	"gopkg.in/yaml.v2"
 )
 
+// Datadog is the global configuration object
```
Comment (pgimalac) [inline pkg/config/setup/config.go:33]: **🥜 nitpick:** ‏tiniest nitpick: could you move this were it was previously, below the const block ? It's just to make the review easier (minimize changes)

File: pkg/config/aliases_darwin.go
[the remark this answers is not in the document]
Comment (pgimalac) [reply pkg/config/aliases_darwin.go:1]: Oh yeah that's true I didn't notice, but in any case the alias itself shouldn't be a var (if you change the alias you will get unexpected results as it won't change the original value)
They are vars on Windows because they are changed in `osinit`, which is only called from `init`, so they can be considered consts outside of the package
So you can still merge the aliases into a single file, and maybe add a comment to explain the above

**Review by pgimalac — APPROVED** [review]
LGTM, thanks for the fixes !

File: pkg/config/aliases_darwin.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+package config
+
+import (
+	pkgconfigsetup "github.com/DataDog/datadog-agent/pkg/config/setup"
+)
+
+// Aliases to setup package
```
Comment (ogaca-dd) [inline pkg/config/aliases_darwin.go:12]: What is the reason of aliases? Why not renaming `config.DefaultSecurityAgentLogFile` to `pkgconfigsetup.DefaultSecurityAgentLogFile`

**Review by ogaca-dd — COMMENTED** [review]
Note: I did a high level review and didn't check that code behaves as before as the PR is too large.

File: pkg/config/aliases_darwin.go (hunk 7 above)
Comment (dineshg13) [reply pkg/config/aliases_darwin.go:12]: We introduced aliases to limit the code changes during the modularization work. We have a follow-up, Jira, to get rid of the aliases. The follow-up PRs should be simple.

**Review by liustanley — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

File: pkg/serverless/trace/install.json [hunk 8]
```diff
@@ -0,0 +1 @@
+{"install_id":"000001f5-952a-21ee-9801-0e0c566f5cb7","install_type":"manual","install_time":1701972019}
```
Comment (purple4reina) [inline pkg/serverless/trace/install.json:1]: These `install.json` files should not be committed.

**Review by purple4reina — CHANGES_REQUESTED** [review]

File: pkg/serverless/trace/install.json (hunk 8 above)
Comment (derekwbrown) [inline pkg/serverless/trace/install.json:1]: is this change intended?

**Review by derekwbrown — APPROVED** [review]

File: pkg/serverless/trace/install.json (hunk 8 above)
Comment (dineshg13) [reply pkg/serverless/trace/install.json]: This is un-intended. Removed the file.

**Review by purple4reina — APPROVED** [review]

File: .github/CODEOWNERS [hunk 9]
```diff
@@ -308,11 +308,12 @@
...
-/pkg/config/system_probe.go               @DataDog/ebpf-platform
-/pkg/config/system_probe_cws.go           @DataDog/agent-security
-/pkg/config/system_probe_cws_notwin.go    @DataDog/agent-security
-/pkg/config/system_probe_cws_windows.go   @DataDog/windows-kernel-integrations
+/pkg/config/setup                            @DataDog/agent-shared-components
+/pkg/config/setup/process*.go                 @DataDog/processes
+/pkg/config/setup/system_probe.go               @DataDog/ebpf-platform
+/pkg/config/setup/system_probe_cws.go           @DataDog/agent-security
+/pkg/config/setup/system_probe_cws_notwin.go    @DataDog/agent-security
+/pkg/config/setup/system_probe_cws_windows.go   @DataDog/windows-kernel-integrations
```
Comment (ahmed-mez) [inline .github/CODEOWNERS:316]: Please update `/pkg/config/apm.go` as well (it's been renamed in this PR).

**Review by kkhor-datadog — COMMENTED** [review]
Mind adding some QA instructions for the changes made to the process agent? Validating that the process agent status output still works correctly should help, as the changes made to `GetProcessAPIAddressPort` are used for the process agent status check: https://github.com/DataDog/datadog-agent/blob/7ea5a00e59584b302396c90a3957b65bb8a99991/pkg/status/status_process_agent.go#L37

The command to check the process agent status is: `/opt/datadog-agent/embedded/bin/process-agent status`

File: .github/CODEOWNERS (hunk 9 above)
Comment (dineshg13) [reply .github/CODEOWNERS:318]: Done

**Comment by dineshg13** [comment]
@kkhor-datadog i added an instruction for process. Let me know if that suffices. Do we have any documentation on checking process-agent status ?

**Comment by dineshg13** [comment]
@kkhor-datadog i have added QA instruction for process agent.

**Review by kkhor-datadog — APPROVED** [review]

**Review by brycekahle — APPROVED** [review]

**Review by FlorentClarret — CHANGES_REQUESTED** [review]
As discussed on Slack, [this](https://github.com/DataDog/datadog-agent/blob/042b151bb5fbd20ba6ab02c2969568ed765a7b75/tasks/utils.py#L166-L170) should probably be updated 🙂

**Review by darccio — APPROVED** [review]

File: tasks/modules.py [hunk 10]
```diff
@@ -188,6 +191,7 @@ def dependency_path(self, agent_version):
     "pkg/util/filesystem": GoModule("pkg/util/filesystem", independent=True),
     "pkg/util/fxutil": GoModule("pkg/util/fxutil", independent=True),
     "pkg/util/buf": GoModule("pkg/util/buf", independent=True),
+    "pkg/util/hostname/validate": GoModule("pkg/util/hostname/validate", independent=True),
```
Comment (GustavoCaso) [inline tasks/modules.py:194]: Are any of the new modules going to be used by Otel? Do we need to use the latest feature flag to avoid being updated in a go version upgrade? 

Here is the PR that introduced the flag https://github.com/DataDog/datadog-agent/pull/21588

cc @pierotibou

Comment (dineshg13) [reply tasks/modules.py:194]: @GustavoCaso we are not yet using the modules in OTel yet. Once we have serializer and logs agent as a module we will be using them.

**Review by FlorentClarret — APPROVED** [review]

**Review by FlorentClarret — APPROVED** [review]
